File: docs/stack/bench-motor/bench.py
# X0 core primitive benchmark - Python. Identical algorithm & seed to bench.js
# Uses arbitrary-precision int as the bitset (the idiomatic fast path in Python)
import sys, time, json

N = int(sys.argv[1]) if len(sys.argv) > 1 else 5
BANK = int(sys.argv[2]) if len(sys.argv) > 2 else 2000
ITERS = int(sys.argv[3]) if len(sys.argv) > 3 else 1000
MASK32 = 0xFFFFFFFF
st = 20260907

def nxt():
    global st
    st ^= (st << 13) & MASK32
    st ^= st >> 17
    st ^= (st << 5) & MASK32
    return st

def rnd(m): return nxt() % m

def fact(k):
    f = 1
    for i in range(2, k + 1): f *= i
    return f

NF = fact(N)
perms = []
for idx in range(NF):
    avail = list(range(N)); rem = idx; p = []
    for i in range(N):
        f = fact(N - 1 - i); j = rem // f; rem %= f
        p.append(avail.pop(j))
    perms.append(p)
M0 = NF * NF
WORDS = (M0 + 31) >> 5

kind = [0]*BANK; a1 = [0]*BANK; b1 = [0]*BANK; b2 = [0]*BANK
for c in range(BANK):
    kind[c] = rnd(5); a1[c] = rnd(N); b1[c] = rnd(N); b2[c] = rnd(N)

t0 = time.perf_counter()
masks = [0]*BANK
pc = [0]*BANK
for c in range(BANK):
    k = kind[c]; s = a1[c]; v1 = b1[c]; v2 = b2[c]
    acc = 0; count = 0
    for m in range(M0):
        pa = perms[m // NF]; pb = perms[m % NF]
        if k == 0:   sat = pa[s] == v1
        elif k == 1: sat = pb[s] == v1
        elif k == 2:
            sat = False
            for i in range(N):
                if pa[i] == v1 and pb[i] == v2: sat = True; break
        elif k == 3: sat = pa[s] != v1
        else:        sat = pa[s] == v1 or pa[s] == v2
        if sat:
            acc |= (1 << m); count += 1
    masks[c] = acc; pc[c] = count
bank_ms = (time.perf_counter() - t0) * 1000

nv_rejected = sum(1 for c in range(BANK) if pc[c] == 0 or pc[c] == M0)

FULL = (1 << M0) - 1
checksum = 0; uniques = 0
t2 = time.perf_counter()
for _ in range(ITERS):
    sel = [rnd(BANK) for _ in range(8)]
    r = FULL
    for i in sel: r &= masks[i]
    total = r.bit_count()
    if total == 1: uniques += 1
    checksum = (checksum + total) & MASK32
    for drop in range(8):
        r = FULL
        for i in range(8):
            if i == drop: continue
            r &= masks[sel[i]]
        checksum = (checksum + r.bit_count()) & MASK32
algebra_ms = (time.perf_counter() - t2) * 1000

print(json.dumps({
    "lang": "python", "n": N, "M0": M0, "bank": BANK, "iters": ITERS, "words": WORDS,
    "bank_ms": round(bank_ms, 1), "algebra_ms": round(algebra_ms, 1),
    "evals": BANK * M0, "mask_bytes": BANK * WORDS * 4,
    "nv_rejected": nv_rejected, "uniques": uniques, "checksum": checksum
}))

File: docs/stack/bench-motor/bench.js
// X0 core primitive benchmark - JS/Node (what TypeScript compiles to)
// Preset "Ancho": n=5, K=3 -> |M0| = (5!)^2 = 14400
// Same algorithm, same seed, same checksums as bench.py / bench.rs / bench.go

const N = parseInt(process.argv[2] || "5", 10);
const BANK = parseInt(process.argv[3] || "2000", 10);
const ITERS = parseInt(process.argv[4] || "1000", 10);
const SEED = 20260907;

// ---- 32-bit PRNG (xorshift32): bit-identical in JS, Python, Rust, Go ----
let st = SEED >>> 0;
function next() {
  st ^= (st << 13) >>> 0; st >>>= 0;
  st ^= st >>> 17;
  st ^= (st << 5) >>> 0; st >>>= 0;
  return st >>> 0;
}
function rnd(m) { return next() % m; }

// ---- permutations via Lehmer decode ----
function fact(k) { let f = 1; for (let i = 2; i <= k; i++) f *= i; return f; }
const NF = fact(N);
const perms = new Uint8Array(NF * N);
{
  const avail = new Array(N);
  for (let idx = 0; idx < NF; idx++) {
    for (let i = 0; i < N; i++) avail[i] = i;
    let rem = idx;
    for (let i = 0; i < N; i++) {
      const f = fact(N - 1 - i);
      const j = Math.floor(rem / f);
      rem = rem % f;
      perms[idx * N + i] = avail[j];
      avail.splice(j, 1);
    }
  }
}
const M0 = NF * NF;              // models: (pi1, pi2)
const WORDS = (M0 + 31) >>> 5;   // uint32 words per mask

// ---- clue bank: T1 (junto), T2 (not junto), T4 (disjunction of 2) ----
// kinds: 0 = SxL, 1 = SxO, 2 = LxO (derived), 3 = negated SxL, 4 = disjunction over L
const kind = new Uint8Array(BANK), a1 = new Uint8Array(BANK),
      b1 = new Uint8Array(BANK), b2 = new Uint8Array(BANK);
for (let c = 0; c < BANK; c++) {
  kind[c] = rnd(5); a1[c] = rnd(N); b1[c] = rnd(N); b2[c] = rnd(N);
}

function popcount32(v) {
  v = v - ((v >>> 1) & 0x55555555);
  v = (v & 0x33333333) + ((v >>> 2) & 0x33333333);
  v = (v + (v >>> 4)) & 0x0f0f0f0f;
  return (Math.imul(v, 0x01010101) >>> 24);
}

// ---- build the bank of masks: the dominant cost of generation ----
const t0 = process.hrtime.bigint();
const masks = new Uint32Array(BANK * WORDS);
const pc = new Int32Array(BANK);
for (let c = 0; c < BANK; c++) {
  const k = kind[c], s = a1[c], v1 = b1[c], v2 = b2[c];
  const base = c * WORDS;
  let count = 0;
  for (let m = 0; m < M0; m++) {
    const pa = ((m / NF) | 0) * N, pb = (m % NF) * N;
    let sat;
    switch (k) {
      case 0: sat = perms[pa + s] === v1; break;                       // junto(s, lugar)
      case 1: sat = perms[pb + s] === v1; break;                       // junto(s, objeto)
      case 2: {                                                         // junto(lugar, objeto)
        sat = false;
        for (let i = 0; i < N; i++) {
          if (perms[pa + i] === v1 && perms[pb + i] === v2) { sat = true; break; }
        }
        break;
      }
      case 3: sat = perms[pa + s] !== v1; break;                       // no junto
      default: sat = perms[pa + s] === v1 || perms[pa + s] === v2;     // disyuncion
    }
    if (sat) { masks[base + (m >>> 5)] |= (1 << (m & 31)); count++; }
  }
  pc[c] = count;
}
const t1 = process.hrtime.bigint();

// ---- NV guard: 0 < popcount(mask) < |M0| ----
let nvRejected = 0;
for (let c = 0; c < BANK; c++) if (pc[c] === 0 || pc[c] === M0) nvRejected++;

// ---- generation loop: residual + NR over sets of 8 clues ----
const scratch = new Uint32Array(WORDS);
const sel = new Int32Array(8);
let checksum = 0, uniques = 0;
const t2 = process.hrtime.bigint();
for (let it = 0; it < ITERS; it++) {
  for (let i = 0; i < 8; i++) sel[i] = rnd(BANK);
  // residuo(S) = AND de mask(c)
  scratch.fill(0xffffffff);
  if ((M0 & 31) !== 0) scratch[WORDS - 1] = (1 << (M0 & 31)) - 1;
  for (let i = 0; i < 8; i++) {
    const base = sel[i] * WORDS;
    for (let w = 0; w < WORDS; w++) scratch[w] &= masks[base + w];
  }
  let total = 0;
  for (let w = 0; w < WORDS; w++) total += popcount32(scratch[w]);
  if (total === 1) uniques++;
  checksum = (checksum + total) >>> 0;
  // NR: quitar cada pista debe dejar >= 2 modelos
  for (let drop = 0; drop < 8; drop++) {
    scratch.fill(0xffffffff);
    if ((M0 & 31) !== 0) scratch[WORDS - 1] = (1 << (M0 & 31)) - 1;
    for (let i = 0; i < 8; i++) {
      if (i === drop) continue;
      const base = sel[i] * WORDS;
      for (let w = 0; w < WORDS; w++) scratch[w] &= masks[base + w];
    }
    let t = 0;
    for (let w = 0; w < WORDS; w++) t += popcount32(scratch[w]);
    checksum = (checksum + t) >>> 0;
  }
}
const t3 = process.hrtime.bigint();

const ms = (a, b) => Number(b - a) / 1e6;
console.log(JSON.stringify({
  lang: "js/node", n: N, M0, bank: BANK, iters: ITERS, words: WORDS,
  bank_ms: +ms(t0, t1).toFixed(1),
  algebra_ms: +ms(t2, t3).toFixed(1),
  evals: BANK * M0,
  mask_bytes: BANK * WORDS * 4,
  nv_rejected: nvRejected, uniques, checksum: checksum >>> 0
}));
